src/token-processor/queue/job-queue.ts

```diff
@@ -0,0 +1,79 @@
+import PQueue from 'p-queue';
+import { PgStore } from '../../pg/pg-store';
+import { DbJob, DbJobStatus } from '../../pg/types';
+import { ENV } from '../../env';
+import { ProcessSmartContractJob } from '../process-smart-contract-job';
+import { ProcessTokenJob } from '../process-token-job';
+
+export enum TokenMetadataProcessingMode {
+  /**
+   * If a recoverable processing error occurs, we'll try again until the max retry attempt is
+   * reached. See `.env`
+   **/
+  default,
+  /** If a recoverable processing error occurs, we'll try again indefinitely. */
+  strict,
+}
+
+/**
+ * Contains a priority queue that organizes all necessary work for contract ingestion and token
+ * metadata processing.
+ */
+export class JobQueue {
```

Comment:
Could we have an explanation of how the job queuing works (perhaps the readme.md, or a code comment)? The queuing system I originally implemented in the API for the old system was a convoluted mess. This looks much cleaner, but I'm still not totally sure how it works. 

Maybe also something like prometheus gauges that show things like "jobs currently being processed", "jobs waiting in queue/memory", "total jobs TODO in the db", "jobs failed/retried (broken down by time)", etc.. could allow for a nice dashboard to easily see what's going on. The current system in the API lacks a lot of visibility around this.

I'm specifically interested in being able to see/notice things like "oh this NFT collection is taking up a lot of queue spaces and resources", and "token XYZ is failing to resolve in an unexpected way and pegging one of the job slots", etc..

Reply:
> Could we have an explanation of how the job queuing works (perhaps the readme.md, or a code comment)? 

Will do!

> Maybe also something like prometheus gauges

Amazing idea, I'll work on this